cloud/validations.go

```diff
@@ -0,0 +1,256 @@
+		"config":            map[string]interface{}{"type": "object"},
+		"regions":           regionsSchema,
+		"region-config":     map[string]interface{}{"type": "object"},
+	},
+	"additionalProperties": false,
+}
+
+var regionsSchema = map[string]interface{}{
+	"type": "object",
+	"additionalProperties": map[string]interface{}{
+		"type": "object",
+		"properties": map[string]interface{}{
+			"endpoint":          map[string]interface{}{"type": "string"},
+			"identity-endpoint": map[string]interface{}{"type": "string"},
+			"storage-endpoint":  map[string]interface{}{"type": "string"},
+		},
+		"additionalProperties": false,
+	},
+}
+
+func validateCloudSet(data []byte) error {
+	var err error
+
+	invalidKeys := validateCloudSetMetaData(data)
+
+	if err != nil {
+		return err
+	}
+
+	formatKeyError := func(invalidKey, similarValidKey string) string {
+		str := fmt.Sprintf("property %s is invalid.", invalidKey)
+		if similarValidKey != "" {
+			str = fmt.Sprintf("%s Perhaps you mean \"%s\".\n", str, similarValidKey)
+		} else {
+			str += "\n"
+		}
+
+		return str
+	}
+
+	msg := ""
+	for k, v := range invalidKeys {
+		msg += formatKeyError(k, v)
+	}
+
+	if msg != "" {
+		err = errors.New(msg)
+	}
+
+	return err
+}
+
+func CloudTags() []string {
+	keys := make(map[string]struct{})
+	collectTags(reflect.TypeOf((*cloud)(nil)), "yaml", []string{"map[string]*cloud.region", "yaml.MapSlice"}, &keys)
+	keyList := make([]string, 0, len(keys))
+	for k := range keys {
+		keyList = append(keyList, k)
+	}
+
+	return keyList
+}
+
+// collectTags returns a set of keys for a specified struct tag. If no tag is
+// specified for a particular field of the argument struct type, then the
+// all-lowercase field name is used as per Go tag conventions. If the tag
+// specified is not the name a conventionally formatted go struct tag, then the
+// results of this function are invalid.
+
+func collectTags(t reflect.Type, tag string, ignoreTypes []string, keys *map[string]struct{}) {
+	switch t.Kind() {
+
+	case reflect.Invalid:
+		panic("invalid kind encountered")
+
+	case reflect.Array, reflect.Slice, reflect.Map, reflect.Ptr:
+		collectTags(t.Elem(), tag, ignoreTypes, keys)
+
+	case reflect.Struct:
+		for i := 0; i < t.NumField(); i++ {
+			field := t.Field(i)
+
+			fieldTag := field.Tag.Get(tag)
+			var fieldTagKey string
+
+			ignoredType := false
+			for _, it := range ignoreTypes {
+				if field.Type.String() == it {
+					ignoredType = true
+					break
+				}
+			}
+
+			if fieldTag == "-" || ignoredType {
+				continue
+			}
+
+			if len(fieldTag) > 0 {
+				fieldTagKey = strings.Split(fieldTag, ",")[0]
+			} else {
+				fieldTagKey = strings.ToLower(field.Name)
+			}
+
+			(*keys)[fieldTagKey] = struct{}{}
+			collectTags(field.Type, tag, ignoreTypes, keys)
+		}
+	}
+}
+
+func validateCloudSetMetaData(data []byte) map[string]string {
+	var body interface{}
+	if err := yaml.Unmarshal(data, &body); err != nil {
+		panic(err)
```

Comment:
Please return the error, don't panic. It's a Juju project policy to avoid panic in general.